openweathermap-ios/openweathermap-ios/Maps/ViewControllers/MapViewController.swift

```diff
@@ -57,12 +63,20 @@ extension MapViewController: MKMapViewDelegate {
     }
     
     func mapView(_ mapView: MKMapView, regionDidChangeAnimated animated: Bool) {
-        updateAnnotation(location: mapView.centerCoordinate)
+        let newCoordinate = mapView.centerCoordinate
+        let distance = newCoordinate.betweenTwoPoints(coordinate: lastCenterCoordinate)
+        if distance > 20_000 {
+            lastCenterCoordinate = newCoordinate
+            weatherManager.fetchWeather(latitude: newCoordinate.latitude, longitude: newCoordinate.longitude)
+        }
         mapView.overlays.forEach { overlay in
             if overlay is MKCircle {
                 mapView.removeOverlay(overlay)
             }
         }
+        if distance < 20_000 {
+            updateAnnotation()
```

Comment:
You should not update annotation until the request get finished.

Reply:
I think this flag `isPin` does not solve the problem. You never set `isPin` to false again, so this line will be called before the current fetch succeeds. 

Trace Table A)
lastCenterCoordinate = 0
isPin = false
regionDidChangeAnimated invoked with map.center = 10k
fetchWeather is not called because distance (10k) is < 20k
updateAnnotation is not called because distance is < 20k and isPin is false

Trace Table 2)
lastCenterCoordinate now is 10k
isPin is still false
regionDidChangeAnimated invoked with map.center = 50k
fetchWeather is called because distance (30k) is > 20k (fetch #1 started) 
updateAnnotation is not called because isPin still false
5 seconds later
fetch #1 succeed turning isPin = true
updateAnnotation was already called in didUpdateWeather method delegate

Trace Table 3)
lastCenterCoordinate now is 50k
*isPin now is true*
regionDidChangeAnimated invoked with map.center = 90k
fetchWeather is called because distance (40k) is > 20k *(fetch #2 started)*
updateAnnotation is called because distance is > 20k and isPin is true *here is the error*


*here is the error*: you are updating an annotation that was already updated in the previous fetching (fetch #1). So, you cannot update any annotation in the same method you start the fetch because the fetch is an async task.

Reply:
I think I understand the problem. I don't know how to solve it, I searched Github to see if there was something similar but I didn't find it.

to stay within the regionDidChangeAnimated func?

```
if distance < 20_000 && isPin {
            updateAnnotation()
        }
```

and changing isPin to false will always request a request

```
if distance > 20_000 {
            isPin = false
            lastCenterCoordinate = newCoordinate
            weatherManager.fetchWeather(latitude: newCoordinate.latitude, longitude: newCoordinate.longitude)
        }
```

Can this be a solution or does it still have the same problem?

Reply:
I don't know what to do to solve this problem. Can you help me?